### ISSUE TITLE
java.lang.IndexOutOfBoundsException: Inconsistency detected. Invalid view holder adapter positionViewHolder

### ISSUE THREAD

**Issue by lostation** [davideas/FlexibleAdapter#278]
Hi Davideas & developers,

I'm facing to a strange problem and I don't know if I do something wrong with the FlexibleAdapter.

To explain the situation : 

In a form, I can open different list of choices from some clicks on edit text view.
To fill the list of choices, I'm using a content provider that gets back a cursor from database fulfilled with some HEADER type and SUBITEM type. Both are stored in a LIST of storage choice model. 

This LIST is then used as input of my implementation of FlexibleAdapter.

At this step, I'm reading all the list content and create ExpandableHeaderItem if type is Header (with sticky feature activated) and SubItem is other... The process of creation is quite simple, see below

``` java
items.clear();   // flush old items
ExpandableHeaderItem currentHeader = null;
for (int i = 0; i < choiceModelList.size(); i++) {
      // HEADER
      if (choiceModelList.get(i).getItemType() == ChoiceModel.GROUP_TYPE) {
                    currentHeader = new ExpandableHeaderItem(choiceModelList.get(i));
                    items.add(currentHeader);
       } else if (choiceModelList.get(i).getItemType() == ChoiceModel.ITEM_TYPE) {
       // ITEM
                    SubItem subItem = new SubItem(choiceModelList.get(i), currentHeader);
                     if (currentHeader != null) {
                           currentHeader.addSubItem(subItem);
                     } else {
                           items.add(subItem);
                     }
        }
}
```
So I'm using ExpandableHeaderItem extending
``` java
AbstractModelItem<ExpandableHeaderItem.ExpandableHeaderViewHolder>
and SubItem extends AbstractModelItem<SubItem.ChildViewHolder>
```
And Sticky is initialized by
``` java
@Override
	public ViewGroup getStickySectionHeadersHolder() {
		FrameLayout frameLayout = new FrameLayout(mRecyclerView.getContext());
        frameLayout.setLayoutParams(new ViewGroup.LayoutParams(
				ViewGroup.LayoutParams.MATCH_PARENT,//or MATCH_PARENT
				ViewGroup.LayoutParams.WRAP_CONTENT));

        //This is important otherwise the Header disappears!
		((ViewGroup) mRecyclerView.getParent()).addView(frameLayout);
		return (ViewGroup) LayoutInflater.from(AproPlanApplication.getAppContext()).inflate(R.layout.sticky_header_layout, frameLayout);
	}
```
On content initialization all seems is perfect, I get my complete list of headers and subitems for each headers. And I can click on header and its content collapsed/uncollapsed -> Working pretty well !!!
The same for sticky behavior, headers are pushing others even if expanded or collapsed.

So all works fine, I choose element and quit the list to open another with another content build exactly as the previous...from cursor, etc ..

Then troubles begin, this time I click on the first or second sticky headers...and boom my app is crashing... But if there's a third header...it collapsed normally.

The crash is due to : 

UncaughtException: java.lang.IndexOutOfBoundsException: Inconsistency detected. Invalid view holder adapter positionViewHolder{d542b3a position=2 id=-1612110282, oldPos=2, pLpos:-1 scrap [attachedScrap] tmpDetached no parent}
 at android.support.v7.widget.RecyclerView$Recycler.validateViewHolderForOffsetPosition (RecyclerView.java:5249)
at android.support.v7.widget.RecyclerView$Recycler.tryGetViewHolderForPositionByDeadline (RecyclerView.java:5431) 
at android.support.v7.widget.RecyclerView$Recycler.getViewForPosition (RecyclerView.java:5392) at android.support.v7.widget.RecyclerView$Recycler.getViewForPosition (RecyclerView.java:5388).... etc 

So it's seems I try to get a header item from my recyclerView that is not available anymore out of boundaries... 

I also see this print on my log appearing sometimes when I scroll the list content with collapsed header content. 

View: isRecyclable decremented below 0: unmatched pair of setIsRecyable() calls for ViewHolder{ade43eb position=3 id=-32702395, oldPos=-1, pLpos:-1 no parent}

I tried with or without Sticky activated.
I tried with recyclerView.setHasFixedSize(true | false);

Do you already have this kind of problem ? Any hint will be appreciated ...

Can you tell me if I need to use some initialization or refresh tips ? 
Manually add setIsRecyclable(true) or setIsRecyclable(false) somewhere...

Thanks in advance.
Lo.

**Comment by davideas** [comment]
@lostation, this might occurs on full list change, so it depends a lot of how you notify the adapter about the change. You can read a lot about this issue in Internet. However, how you make the update?

**Comment by inorichi** [comment]
I think I had the same issue. You probably forgot to update your adapter with a COPY of the list. If you update the adapter without creating a new list, this will happen.

**Comment by lostation** [comment]
Thanks for answer

So before any update adapter content : I do
if (items != null) {
            LOGGER.e(TAG, "Remove all content");
            items.clear();
            notifyDataSetChanged();
}
then     items = new ArrayList<>();

Then in a FOR loop  I add some  items.add(currentHeader);  and  currentHeader.addSubItem(subItem);

Then call   updateDataSet(items, animate);

 @Override
    public void updateDataSet(List<AbstractFlexibleItem> items, boolean animate) {
        //Overwrite the list and fully notify the change, pass false to not animate changes.
        //Watch out! The original list must be a copy

OK then do a copy
        List<AbstractFlexibleItem> listToAdapter = new ArrayList<>(items);
        notifyDataSetChanged();

then pass it to adapter
        super.updateDataSet(listToAdapter, animate);
}

I don't know what to do more...on update.

I suppose In case of huge list, from a cursor to a first list of item then copy content again in a new list with the same content... -> explosion of memory

**Comment by lostation** [comment]
I use     compile 'eu.davidea:flexible-adapter:5.0.0-rc1'

**Comment by lostation** [comment]
Like I said, I use several list with different content that use the flexibleAdapter and the click on item or header works really well... but for some header on the same list ... crash

**Comment by davideas** [comment]
@lostation, the copy as you know is only by reference for a list or set.
I see few things not ok:
- Normally, you have your own list of the database and the adapter list, as explained the adapter needs to modify the list when expanding, searching, it deletes and adds items as you command. So keep in mind this architecture. Almost all advanced features cannot work if the list is linked to a cursor.
- The `updateDataSet()` with `animate=false` already invokes `notifyDataSetChanged()`, since the synchronization is already done by the adapter! So, you can forget to call all the methods `notify**` when you use FlexibleAdapter.
- Since the synchronization is done by the adapter, there are 2 ways to animate the changes:
   - items have unique ids (stableIds): RecyclerView can animate changes even with notifyDataSetChanges();
   - more granular way is to enable animate=true where all single modifications are notified.

Big list is also supported by the internal synchronization and it's faster than `DiffUtil`.

**Comment by lostation** [comment]
I found my mistake. I was not a question of copy of dataset neither notify* methods...
In my dataset's content sometimes I used same value as title in header views and in my abstract model I was using a wrong implementation of "id". I used to get the id based on the title...but sometime this value was the same than another item...and I was hashing this value twice on the same value so recycler was fighting with problematic of id. So to resolve that issue... I just added 

public AbstractModelItem() {
	this.id = UUID.randomUUID().toString();
}

so I'm ensure that my id as really unique !!!

Now it's perfectly working in folding/unfolding header content.
